### PR TITLE
Convert to cmake and container environment.\n\nFixes #16

### PR DESCRIPTION
**Pull request by liangbin-pub**
Convert to cMake/cTest/cPack build system for next stage

Making adjustment to deployment scripts to separate local test and live deployment environments

Update getting start document to reflect the changes

### CODE REVIEW

File: deploy/build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,70 @@
+FROM ubuntu:18.04
```
Comment (phudtran) [inline deploy/build/Dockerfile]: This container takes quite a long time to build.
If we are simply using it to build our binaries, maybe we can push the image to some registry and just pull the image on demand.

We can then just keep the Dockerfile definition here as documentation, and make changes to it as necessary.

File: deploy/kind/Dockerfile [hunk 2]
```diff
@@ -20,9 +19,9 @@
...
-DOCKER_ACC="fwnetworking"
-
-source deploy/install/create_crds.sh $CWD
-source deploy/install/deploy_operator.sh $CWD $DEPLOY $DOCKER_ACC
+FROM localhost:5000/zetanode:latest
+COPY . /var/zeta/
+RUN pip3 install /var/zeta/deploy
+RUN ln -snf /var/zeta/build/bin /trn_bin
+RUN ln -snf /var/zeta/build/xdp /trn_xdp
+RUN ln -snf /var/zeta/build/tests/zeta.config /etc/zeta.config
```
Comment (phudtran) [inline deploy/kind/Dockerfile]: Nitpick, new line at the end of file.

File: src/dmn/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+add_dependencies(test_dmn libbpf rpcgen)
+# Add test coverage compiler flags
+target_compile_options(test_dmn PRIVATE -fprofile-arcs -ftest-coverage)
+# Add CMocka unit test and cJSON to linker flags
+target_link_libraries(test_dmn extern -lgcov --coverage -lcmocka ${DMN_MOCKS})
+set_target_properties(test_dmn PROPERTIES
+                      RUNTIME_OUTPUT_DIRECTORY_DEBUG ${CMAKE_BINARY_DIR}/tests
+                      RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_BINARY_DIR}/tests
+)
+add_test(NAME test_dmn COMMAND test_dmn)
```
Comment (phudtran) [inline src/dmn/CMakeLists.txt]: new line

File: build.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	--cap-add=SYS_PTRACE \
+	--security-opt seccomp=unconfined \
+	--name zb \
+	--workdir /mnt/host/code \
+	zeta_build:latest /bin/bash
+docker start zb
+
+# Build zeta
+echo "--- building zeta ---"
+docker exec zb bash -c "cd /mnt/host/code/build && cmake -H/mnt/host/code -B. -DCMAKE_BUILD_TYPE=$FLAVOR -DARCH=$arch -DBUILD_TESTING=On && make $args && ctest -V"
```
Comment (phudtran) [inline build.sh]: new line

File: docs/user/getting_started.md [hunk 5]
```diff
@@ -48,16 +55,34 @@ If you are testing out Zeta with Kind, a script has been included to setup a loc
...
 This script does the following:
 
-* Create a multi-node local Kubernetes cluster with Kind.
+* Create a multi-node (default is 3) local Kubernetes cluster with Kind.
 * Build the Kind-Node, and Zeta-Operator docker images
 * Apply all of the Zeta [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
 * Deploy the Zeta [Operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).
 
-### Linux Kernel Update
+### Deploy to Remotr K8S Cluster
```
Comment (phudtran) [inline docs/user/getting_started.md]: "Remotr" -> "Remote"

**Review by phudtran — CHANGES_REQUESTED** [review]
I ran into this error with the build container when running `./build.sh debug `after cloning.
Did I miss something?
```
Step 8/9 : RUN addgroup --gid $GROUP_ID user &&     adduser --disabled-password --gecos '' --uid $USER_ID --gid $GROUP_ID user
 ---> Running in 881f2cd29cfa
addgroup: The GID `0' is already in use.
The command '/bin/sh -c addgroup --gid $GROUP_ID user &&     adduser --disabled-password --gecos '' --uid $USER_ID --gid $GROUP_ID user' returned a non-zero code: 1
Error: No such container: zb
Unable to find image 'zeta_build:latest' locally
Error response from daemon: pull access denied for zeta_build, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
Error response from daemon: No such container: zb
Error: failed to start containers: zb
--- building zeta ---
Error: No such container: zb
```

File: deploy/build/Dockerfile (hunk 1 above)
Comment (liangbin-pub) [reply deploy/build/Dockerfile]: yes, we can have a base image in public registry
The 1st error is caused by your user id not in docker group:
sudo usermod -aG docker $USER
newgrp docker

File: deploy/kind/Dockerfile (hunk 2 above)
Comment (liangbin-pub) [reply deploy/kind/Dockerfile]: Taken

File: src/dmn/CMakeLists.txt (hunk 3 above)
Comment (liangbin-pub) [reply src/dmn/CMakeLists.txt]: taken

File: build.sh (hunk 4 above)
Comment (liangbin-pub) [reply build.sh]: taken

File: docs/user/getting_started.md (hunk 5 above)
Comment (liangbin-pub) [reply docs/user/getting_started.md]: :) what can I say, must accept

**Comment by liangbin-pub** [comment]
> I ran into this error with the build container when running `./build.sh debug `after cloning.
> Did I miss something?
> 
> ```
> Step 8/9 : RUN addgroup --gid $GROUP_ID user &&     adduser --disabled-password --gecos '' --uid $USER_ID --gid $GROUP_ID user
>  ---> Running in 881f2cd29cfa
> addgroup: The GID `0' is already in use.
> The command '/bin/sh -c addgroup --gid $GROUP_ID user &&     adduser --disabled-password --gecos '' --uid $USER_ID --gid $GROUP_ID user' returned a non-zero code: 1
> Error: No such container: zb
> Unable to find image 'zeta_build:latest' locally
> Error response from daemon: pull access denied for zeta_build, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
> Error response from daemon: No such container: zb
> Error: failed to start containers: zb
> --- building zeta ---
> Error: No such container: zb
> ```

Caused by Running as root, latest change enables running as root but this is not recommended

**Review by phudtran — APPROVED** [review]
Thanks Bin, LGTM

**Pull request merged**